broker.go

```diff
@@ -0,0 +1,604 @@
+}
+
+// Close shuts down the broker.
+func (lb *LegBroker) Close() error {
+	lb.unsubEvents()
+	lb.psub.Cancel()
+	lb.cancelBrkr()
+
+	// Dismiss any event readers.
+	lb.outEventsMutex.Lock()
+	for _, ch := range lb.outEventsChans {
+		close(ch)
+	}
+	lb.outEventsChans = nil
+	lb.outEventsMutex.Unlock()
+
+	// Dismiss any handlers waiting completion of sync.
+	lb.syncDoneMutex.Lock()
+	if len(lb.syncDoneChans) != 0 {
+		log.Warnf("Closing broker with %d syncs in progress", len(lb.syncDoneChans))
+	}
+	for _, ch := range lb.syncDoneChans {
+		close(ch)
+	}
+	lb.syncDoneChans = nil
+	lb.syncDoneMutex.Unlock()
+
+	// Shutdown datatransfer.
+	lb.cancelAll()
+	return nil
+}
+
+// OnChange creates a channel that receives change notifications, and adds that
+// channel to the list of notification channels.
+//
+// Calling the returned cancel function removed the notification channel from
+// those the receive change notifications, and closes the channel to allow any
+// waiting goroutines to stop waiting on the channel.  For this reason, channel
+// readers should check for channel closure.
+func (lb *LegBroker) OnChange() (<-chan ChangeEvent, context.CancelFunc) {
+	// Channel is buffered to prevent distribute() from blocking if a reader is
+	// not reading the channel immediately.
+	ch := make(chan ChangeEvent, 1)
+	lb.outEventsMutex.Lock()
+	defer lb.outEventsMutex.Unlock()
+
+	lb.outEventsChans = append(lb.outEventsChans, ch)
+	cncl := func() {
+		lb.outEventsMutex.Lock()
+		defer lb.outEventsMutex.Unlock()
+		for i, ca := range lb.outEventsChans {
+			if ca == ch {
+				lb.outEventsChans[i] = lb.outEventsChans[len(lb.outEventsChans)-1]
+				lb.outEventsChans[len(lb.outEventsChans)-1] = nil
+				lb.outEventsChans = lb.outEventsChans[:len(lb.outEventsChans)-1]
+				close(ch)
+				break
+			}
+		}
+	}
+	return ch, cncl
+}
+
+// Sync performs a one-off explicit sync with the given peer for a specific CID and updates the latest
+// synced link to it.
+//
+// The Context passed in controls the lifetime of the background sync process
+// and not the call to sync, so make sure the context has an appropriate
+// deadline, if any.
+//
+// A change event channel is returned to allow the caller to wait for this particular sync
+// to complete.  Any OnChange watchers will also get a ChangeEvent when the sync succeeds.
+//
+// It is the responsibility of the caller to make sure the given CID appears after the latest sync
+// in order to avid re-syncing of content that may have previously been synced.
+//
+// The selector sequence, selSec, can optionally be specified to customize the
+// selection sequence during traversal.  If unspecified, the subscriber's
+// default selector sequence is used.
+//
+// Note that the selector sequence is wrapped with a selector logic that will stop traversal when
+// the latest synced link is reached. Therefore, it must only specify the selection sequence itself.
+//
+// See: ExploreRecursiveWithStopNode.
```

Comment:
Yes to the first part.  We want to walk back as far as is reasonable, possibly to the beginning of a provider's records, to get all content.  A new indexer may come online at any time and will want to get content that has already been published.  Also handles missed messages.

The second part is because that is how we make IPLD/graphsync walk a DAG (a chain in this case).